### PR TITLE
Fix unknown prop warnings, phase 2

### PR DESCRIPTION
**Pull request by taion**

### CODE REVIEW

**Comment by taion** [comment]
This is good to go now. I cleaned up a bit of the state management in `<Nav>` and `<TabPane>` as well.

**Comment by taion** [comment]
Just the overlays left!

File: test/TabContainerSpec.js [hunk 1]
```diff
@@ -308,6 +314,46 @@ describe('<TabContainer>', () => {
     instance.single('[eventKey=0]').single('.active');
   });
 
+  [
+    [[1, 2], [2, 1]],
+    [[2, 1], [1, 2]],
+  ].forEach(([order1, order2]) => {
+    it('should handle event key swaps', () => {
```
Comment (taion) [inline test/TabContainerSpec.js]: @jquense 

Never mind, I found another tab animation edge case (the logic in the previous commit would drop the animation if this happened... the logic on master would instead fail the duplicate `eventKey` invariant)

basically doing this type of imperative animation end callback handling in React is the worst thing ever. I don't think even not using context would simplify this appreciably.

**Comment by Nishchit14** [comment]
Any estimation for next release  please?, this warning messages feels like a mad dragon in the heaven while consoling.

**Comment by taion** [comment]
Use the documented warning filtering functionality

File: src/Carousel.js [hunk 2]
```diff
@@ -122,214 +123,224 @@ let Carousel = React.createClass({
...
-  },
-
-  pause() {
-    this.isPaused = true;
-    clearTimeout(this.timeout);
-  },
+    this.select(index, e, 'prev');
+  }
 
-  play() {
```
Comment (jquense) [inline src/Carousel.js]: these are missing...my hunch is that they are public methods.

Comment (taion) [inline src/Carousel.js]: You think so? They're not documented anywhere. They'll break as soon as we move this to use uncontrollable anyway. IMO if someone complains we'd be better off just adding a prop to control this.

File: src/DropdownButton.js [hunk 3]
```diff
@@ -1,57 +1,57 @@
...
-          disabled={disabled}
+          bsClass={prefix(props, 'toggle')}
+          bsSize={bsSize}
+          bsStyle={bsStyle}
         >
           {title}
         </Dropdown.Toggle>
-        <Dropdown.Menu>
+
+        <Dropdown.Menu bsClass={prefix(props, 'menu')}>
```
Comment (jquense) [inline src/DropdownButton.js]: is this how we treat other subcomponents with bsClass?

File: src/DropdownMenu.js [hunk 4]
```diff
@@ -2,22 +2,32 @@ import classNames from 'classnames';
...
+const defaultProps = {
+  bsRole: 'menu',
+  pullRight: false,
+};
 
 class DropdownMenu extends React.Component {
   constructor(props) {
     super(props);
 
-    this.focusNext = this.focusNext.bind(this);
```
Comment (jquense) [inline src/DropdownMenu.js]: these are "public" in that they are called by other components (dropdown) we should bind them as a defensive measure I think.

File: src/DropdownToggle.js [hunk 5]
```diff
@@ -3,56 +3,54 @@ import classNames from 'classnames';
...
+};
 
 class DropdownToggle extends React.Component {
   render() {
     const {
-      noCaret,
-      useAnchor,
-      bsClass: bootstrapClass,
-      ...props
+      noCaret, open, useAnchor, bsClass, className, children, ...props,
```
Comment (jquense) [inline src/DropdownToggle.js]: in general I think if we are gonna push properties to a new line we should jsut format them all on a new line...I find this style harder to read personally, not sure why.

File: src/DropdownButton.js (hunk 3 above)
Comment (taion) [inline src/DropdownButton.js]: No – this is new. I think this is the right way to handle it. Previously `bsClass` would get passed down from `<Dropdown>`, which made it impossible to directly set a custom `bsClass` on the subcomponents.

This just seemed to be the most intuitive way to go – if you set a `bsClass` on `<DropdownButton>`, it propagates through to the subcomponents.

File: src/DropdownToggle.js [hunk 6]
```diff
@@ -3,56 +3,54 @@ import classNames from 'classnames';
...
-        className={classNames(classes, this.props.className)}
         role="button"
+        className={classNames(className, bsClass)}
         aria-haspopup
-        aria-expanded={this.props.open}>
-        {this.props.children || this.props.title}{caret}
+        aria-expanded={open}
+      >
+        {children || props.title}
+        {useCaret && ' '}
```
Comment (jquense) [inline src/DropdownToggle.js]: should we be making this assumption for a user? I think the extra space probably isn't our job

File: src/Nav.js [hunk 7]
```diff
@@ -5,165 +5,262 @@ import ReactDOM from 'react-dom';
...
 
-  componentDidUpdate() {
-    if (this._needsRefocus) {
-      let ul = this.refs.ul && ReactDOM.findDOMNode(this.refs.ul);
-      let tabs = ul ? ul.children || [] : [];
-      let tabIdx = this.eventKeys().indexOf(this.getActiveKey());
+// TODO: This `bsStyle` is very unlike the others. Should we rename it?
 
-      this._needsRefocus = false;
+// TODO: `pullRight` and `pullLeft` don't render right outside of `navbar`.
```
Comment (jquense) [inline src/Nav.js]: not sure these TODO's are gonna be actionable here, maybe a new issue it better.

on them tho, I think the ship has probably sailed for renaming the style prob (maybe not)

File: src/DropdownToggle.js (hunk 6 above)
Comment (taion) [inline src/DropdownToggle.js]: This was what the old logic did. I'm just writing it this way to take advantage of the `span` elision in React 15.

File: src/Nav.js (hunk 7 above)
Comment (jquense) [inline src/Nav.js]: Not sure that note is right about the navbar...they are simply given the pull-right/left class in the non-navbar instance, there isn't actually an expectation for how that should look in those cases, but i've found them useful regardless.

File: src/DropdownMenu.js (hunk 4 above)
Comment (taion) [inline src/DropdownMenu.js]: DropdownMenu calls it as `this.menu.focusNext()` though, which doesn't require binding. It's a pretty standard caveat with class-like objects in JS, so I figure why make it any different here?

File: src/DropdownButton.js (hunk 3 above)
Comment (jquense) [inline src/DropdownButton.js]: It also effectively makes setting bsClass on the subcomponents (or changing the defaults) a noop though right? Since Dropdown will always have a default for this.

Comment (taion) [inline src/DropdownButton.js]: I took out the logic for `<Dropdown>` itself forwarding through `bsClass`, since it seemed unnecessary.

**Comment by jquense** [comment]
as far as i'm possible, it looks good to me

**Comment by menems** [comment]
Hello, is there a possibility to merge this pr on 0.29.6 ? or a way to use next branch directily with npm ? maybe publish a next tag on npm ?

thanks for the works
